### PR TITLE
handling for GPU, OCR, WIP-File Prediction and various bug fixes

### PR DESCRIPTION
**Pull request by C-K-Loan**
Signed-off-by: Christian Kasim Loan <christian.kasim.loan@gmail.com>

### What changes are proposed in this pull request?

-  Various bug fixes. In particular removing usage of MLFLOWDBFS. It's incredibly buggy and has caused numerous random bugs on Databricks infrastructure. I send a large bug report to databricks on a seperate e-mail thread about this
- support for loading OCR and GPU jars
- Support for parameterization of .predict()
- WIP support for file based predictions
- various bugfixes
- official docs https://nlp.johnsnowlabs.com/docs/en/jsl/databricks-utils

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/johnsnowlabs/__init__.py [hunk 1]
```diff
@@ -161,20 +201,21 @@ def get_default_conda_env():
...
+        sample_input=None,
+        registered_model_name=None,
+        signature: ModelSignature = None,
+        input_example: ModelInputExample = None,
+        await_registration_for=DEFAULT_AWAIT_MAX_SLEEP_SECONDS,
+        pip_requirements=None,
+        extra_pip_requirements=None,
+        metadata=None,
+        store_license=False,
+        gpu=False,
```
Comment (harupy) [inline mlflow/johnsnowlabs/__init__.py:218]: Can we revert unrelated changes?

File: mlflow/johnsnowlabs/__init__.py [hunk 2]
```diff
@@ -807,22 +875,88 @@ def _fetch_deps_from_path(local_model_path):
 
 
 def _unpack_and_save_model(spark_model, dst):
+    print(f'Saving model to {dst}')
```
Comment (harupy) [inline mlflow/johnsnowlabs/__init__.py:878]: print should not be used. can you remove it?

File: mlflow/johnsnowlabs/__init__.py [hunk 3]
```diff
@@ -807,22 +875,88 @@ def _fetch_deps_from_path(local_model_path):
...
+def _maybe_unpack_and_save_model(spark_model, dst):
+    try:
+        _unpack_and_save_model(spark_model, dst)
+        return True
+    except:
+        return False
+
+
+import pandas as pd
+from pydantic import BaseModel
```
Comment (harupy) [inline mlflow/johnsnowlabs/__init__.py:918]: `pydantic` is not an mlflow's dependency. can we use dataclass or namedtuple?

File: mlflow/johnsnowlabs/__init__.py [hunk 4]
```diff
@@ -590,7 +653,7 @@ def save_model(
     # on a Databricks cluster and the URI is schemeless (e.g. looks like a filesystem absolute path
     # like "/my-directory")
     copying_from_dbfs = is_valid_dbfs_uri(tmp_path) or (
-        databricks_utils.is_in_cluster() and posixpath.abspath(tmp_path) == tmp_path
+            databricks_utils.is_in_cluster() and posixpath.abspath(tmp_path) == tmp_path
```
Comment (harupy) [inline mlflow/johnsnowlabs/__init__.py:656]: unnecessary change.

**Comment by harupy** [comment]
@C-K-Loan can you split this PR? A PR that attempts to solve multiple problems is harder to review.

**Comment by C-K-Loan** [comment]
> @C-K-Loan can you split this PR? A PR that attempts to solve multiple problems is harder to review.

Ok let me do that

File: mlflow/johnsnowlabs/__init__.py (hunk 3 above)
Comment (C-K-Loan) [reply mlflow/johnsnowlabs/__init__.py:918]: @harupy  pydantic is a johnsnowlabs dependency. Any user of this flavor must have it installed anyways. 
If they manage to import mlflow.johnsnowlabs, they must have johnsnowlabs installed.
If johnsnowlabs is installed, there must be pydantic in the env, so I think its ok. 
What do you think?

**Comment by C-K-Loan** [comment]
@harupy 
Closing this in favor of these 4 clean branches, please review 
- [remove mlflow dbfs  ](https://github.com/mlflow/mlflow/pull/9872)
- [paramerization](https://github.com/mlflow/mlflow/pull/10596)
- [GPU Jar access](https://github.com/mlflow/mlflow/pull/10597/)
- [Visual Jar  + minor refactor](https://github.com/mlflow/mlflow/pull/10603)
